doc/developer/design/20250203_cluster_reliability.md

```diff
@@ -0,0 +1,224 @@
+unsure, reach out to your manager for help.
+
+Remember to document any dependencies that may need to break or change as a
+result of this work.
+-->
+
+User's needs aren't uniform; a solution working for one customer might be too limiting for another.
+We propose a set of mechanisms that offer a spectrum of solutions:
+* We add a RBAC permission `CREATE DATAFLOW`, which is required to create dataflows.
+  If the permission is absent, select and subscribe queries can only read from existing objects.
+* We add a RBAC permission `READ EXTERNAL`, which is required to read data from external storage (persist).
+  The permission controls both dataflows reading from external sources, and peeks that directly read from external storage.
+* We implement a mechanism to limit the resources of a dataflow serving a query with a probability.
+
+### `CREATE DATAFLOW` permission
+
+The `CREATE DATAFLOW` permission controls whether a query can render its own dataflow.
+Without the permission, Materialize will reject a query that it cannot otherwise fulfill.
+The permission applies to a cluster.
+
+Implementing this requires additional plumbing.
+At the moment, we evaluate the RBAC permissions before sequencing a plan.
+However, we only learn during sequencing whether an object needs to render a dataflow.
+
+### `READ EXTERNAL` permission
+
+The `READ EXTERNAL` permission controls whether a query can read from external storage.
+If a query can only be fulfilled by reading data from external storage, Materialize checks that the issuer has the permission on the target cluster.
+
+The permission would apply to all objects, i.e., indexes, materialized views, and queries.
+Similar considerations as with `CREATE DATAFLOW` regarding the implementation apply here, too.
+
+### Limiting a query's heap size
+
+Measuring heap size usage is inherently difficult as we cannot attribute all memory usage to a single query.
+Any operator can allocate and deallocate memory, and we do not want to (or cannot) instrument all operators.
+We can only measure instrumented dataflow operators, and with a time delay, which renders all solutions approximate.
+
+A solution needs to limit the heap size usage of a query to a configurable value, and enforce it with high probability.
+
+To measure the heap size of a query, we hook into the arrangement heap size infrastructure.
+It already provides the system with a view of the heap size of each arrangement and its merge batcher.
+We can reduce the data for each dataflow, and join it with a per-dataflow limit to obtain the current heap size of a query.
+Reducing the value yields the absolute heap size, which we can apply a threshold to detect limits exceeding their allocation.
+
+We then forward the information to the controller, which can decide on how to handle dataflows with excessive resource utilization.
+It can decide to terminate and fail a query, or let its execution continue.
```

Comment:
Feels like we can surface this exact alert and have separately configurable actions?